### ISSUE TITLE
KFServing dynamic model updates

### ISSUE THREAD

**Issue by bestowjay** [kubeflow/kfserving#772]
/kind feature

**Describe the solution you'd like**
KFServing is (right now) loosely based on Tensorflow-Serving. One great thing that I like about TFServing is that it has the ability to use gcs buckets as the `model_base_path=gs://bucket/model`. When this form is used, TFServing is aware when I push a new model version to the bucket path. This uses an integer subdirectory, e.g. `gs://bucket/model/1` then I tweak a model and save a new version to `gs://bucket/model/2`. TFServing picks up this change to the bucket and syncs the new version for serving as well. It even leaves the old version up so you can use the REST api to query either always the latest version (by omitting the version from the url) or specifying a specific version of the model to use.

As KFServing stands, it's unclear to me how I go about easily updating a model once I've deployed it. Perhaps I am missing something about deploying models, however if we could reach this point of auto-updating. My kubeflow pipelines component only needs to auto-increment a build integer and deploy to a gcs bucket to automatically serve my updated model.

**Anything else you would like to add:**
Let me know if there's any other information you need about this feature, or some other way I can help.

**Comment by ellis-bigelow** [comment]
Thank for you this issue! It's making me fundamentally rethink how we do this. I'd really like to explore this idea. We original set out with the idea of "hermetic infrastructure", where the servers are only intended to point to a single model. If you wanted an upgrade, you should do a canary rollout, meaning config changes at the k8s level. Like you say, this is high touch, and may be beyond what data scientists want to do or even have permission to do.

That rollout pattern isn't the path that model servers have gone down over the last few years, but they weren't developing them in a cloud native world. I think there is a big question around hermeticity: is the infrastructure hermetic, or is the infrastructure and the application data (i.e. models) hermetic.

Currently, our storage initializer snapshots the data for each pod. So unless you restart the pods, you won't get the data. If we wanted to live update an existing inferenceservice, we'd need to continually poll and update the data in a sidecar. There are some challenges here around caching, disk space, etc, so we'd need to be careful. This also somewhat defeats the purpose of our canarying mechanism. You could potentially repurpose it to canary infrastructure differences (i.e. resources and explainer methods) -- we'd need to explore that path and the tradeoffs.

Thoughts @rakelkar @cliveseldon @yuzisun @animeshsingh?

**Comment by cliveseldon** [comment]
Two quick comments:

 * For the popular "GitOps" view the resource should change in source control in some way. This allows for easy auditing and rollbacks.
 * For canarying I think you would want some live testing of the new model which current servers don't support so you would lose this if dynamic updates were allowed unless canarying was part of the model server's repsonsibility. 

The above doesn't mean the simplicity of updating to a new model could not be made much simpler.

**Comment by bestowjay** [comment]
This feature request comes from a sort of interesting case too. We're a small team trying to continuously refine models. We don't have a lot of engineering power to keep developing infrastructure components, so we're kind of starting from where we can get nice fast gains by having pipelines that exist already serve their models into KFServing to abstract the whole serving infrastructure. While we're standing this up, we're moving over to kubeflow completely. This has given rise to this idea that we could allow the less-engineering scientists to keep moving along quickly knowing that they can easily deploy models that will be reachable at an endpoint after they've passed their own testing.

If the canarying mechanism can be used to do a switchover of a new model version at the main endpoint, I think that could even solve this particular use case. We were intrigued by the idea of building a pipeline that did all of the training, validation, some testing, then just "stick it in a bucket" and the serving infrastructure would reconcile the new model into a REST/gRPC endpoint. In looking at how the initContainer stuff happens though, I realize this isn't really an easy thing to even accomplish what with the decoupling of the data infrastructure from the serving infrastructure.

Restarting the pods is essentially what I'm doing, too. Which isn't too difficult, it's just that we now manage a portion of the serving infrastructure.

Thanks for taking the time to look at this request.

**Comment by ellis-bigelow** [comment]
Another option would be to create a "Model" CRD that contains the model's artifacts. If we decoupled InferenceServices (things that serve models) from Models themselves, we might be able to achieve a similar gitops to what you're talking about. I can imagine a storage trigger that when a new folder is added, it goes and creates a new K8s Model resource.

**Comment by yuzisun** [comment]
@bestowjay Maybe I am missing something here, assuming you are using kubeflow pipeline, for a low touch workflow each time you run the pipeline it would produce a new model and saved on storage with a new version, can you actually add KFServing as last step to reconcile the inference service each time a new model is produced? You could parameterize the default storage uri with [KFServing component](https://github.com/kubeflow/kfserving/blob/master/docs/samples/pipelines/sample-tf-pipeline.py#L28) and feed into it from previous steps, so inference service can automatically restart and load the new model.

**Comment by animeshsingh** [comment]
@bestowjay
>> My kubeflow pipelines component only needs to auto-increment a build integer and deploy to a gcs bucket to automatically serve my updated model.

This is super convenient, and if you are using KFP, maybe can addressed by a combination of KFServing + KFP, by adding more specialized functionalities in KFServing Component.
https://github.com/kubeflow/pipelines/blob/master/components/kubeflow/kfserving/src/kfservingdeployer.py

If you this can work, we can discuss the details of the KFServing component extension

@ellis-bigelow 
>> Currently, our storage initializer snapshots the data for each pod. So unless you restart the pods, you won't get the data. If we wanted to live update an existing inferenceservice, we'd need to continually poll and update the data in a sidecar. 

Is there documentation on storage initializer with diagram et all? Also can this be triggered, instead of polling if needed, and in that case an external client like KFP can trigger it.

**Comment by bestowjay** [comment]
> @bestowjay Maybe I am missing something here, assuming you are using kubeflow pipeline, for a low touch workflow each time you run the pipeline it would produce a new model and saved on storage with a new version, can you actually add KFServing as last step to reconcile the inference service each time a new model is produced? You could parameterize the default storage uri with [KFServing component](https://github.com/kubeflow/kfserving/blob/master/docs/samples/pipelines/sample-tf-pipeline.py#L28) and feed into it from previous steps, so inference service can automatically restart and load the new model.

Yeah, I can certainly take a look at trying to update the InferenceService, but it's kind of the same issue with just killing the pod. I have to modify the cluster itself. I would like to avoid giving k8s access to those users that would not need it if I can, but I will use this solution in the meantime if this works. I'll get a chance to try later this week.

> This is super convenient, and if you are using KFP, maybe can addressed by a combination of KFServing + KFP, by adding more specialized functionalities in KFServing Component.

We don't have KFP just yet, but that is our end goal. I'll try to look at using the SDK to do this as well through the KF component, this is kind of the same thing, right? where we're altering the k8s resource?

**Comment by vitobotta** [comment]
Are there any known workarounds for this? I would like the endpoint that serves the model to automatically  pick up new versions.

**Comment by hsinhoyeh** [comment]
GCS[1] and S3[2] have featured to allow you to get notifications when the bucket data got changed. I think this would be a great feature (and make model updates easier) to integrate with Kserve to handle dynamic model updates. we are more than happy to be able to contribute to this if possible.

[1] https://cloud.google.com/storage/docs/object-change-notification
[2] https://docs.aws.amazon.com/AmazonS3/latest/userguide/enable-event-notifications.html
